kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -484,28 +528,164 @@ private static void validateFieldCompatibility(StructField existingField, Struct
-    List<Tuple2<List<String>, StructField>> columnPathToStructField =
-        filterRecursively(
-            schema,
-            true /* recurseIntoMapOrArrayElements */,
-            false /* stopOnFirstMatch */,
-            sf -> true);
-    Map<Integer, StructField> columnIdToField = new HashMap<>();
-    for (Tuple2<List<String>, StructField> pathAndField : columnPathToStructField) {
-      StructField field = pathAndField._2;
-      checkArgument(hasColumnId(field), "Field %s is missing column id", field.getName());
-      checkArgument(hasPhysicalName(field), "Field %s is missing physical name", field.getName());
+  /**
+   * Returns a mapping of id to field wrappers containing the field, its id, its parent id and a
+   * full column path. In the case of icebergCompatV2, for map key/value and array elements, the
+   * parent id is the field ID of the map or array struct field, and the ID will be extracted from
+   * nested column metadata.
+   *
+   * <p>For other cases, map keys/values and array elements, where these elements are not structs
+   * will not be included in this mapping since those will not have IDs. When the elements are
+   * structs, the parent ID will be the nearest containing struct field.
+   *
+   * @param schema
+   * @param icebergCompatV2
+   * @return
+   */
+  static Map<Integer, StructFieldWithIds> fieldsById(StructType schema, boolean icebergCompatV2) {
+    Map<Integer, StructFieldWithIds> fieldsById = new HashMap<>();
+    for (StructField field : schema.fields()) {
+      validateFieldHasIdAndName(field);
       int columnId = getColumnId(field);
       checkArgument(
-          !columnIdToField.containsKey(columnId),
-          "Field %s with id %d already exists",
+          !fieldsById.containsKey(columnId),
+          "Field %s with id %s already exists",
           field.getName(),
           columnId);
-      columnIdToField.put(columnId, field);
+      fieldsById.put(
+          columnId, new StructFieldWithIds(field, getColumnId(field), null, field.getName()));
+      idToFields(field, icebergCompatV2, fieldsById, null, getPhysicalName(field));
+    }
+
+    return fieldsById;
+  }
+
+  /**
+   * StructFieldWrapper wraps a StructField along with its field ID, a parent ID, and the full
+   * column path.
+   */
+  static class StructFieldWithIds implements Supplier<StructField> {
+    private final StructField field;
+    private final int id;
+    private final Integer parentId;
+    private final String columnPath;
+
+    StructFieldWithIds(StructField field, int id, Integer parentId, String columnPath) {
+      this.field = field;
+      this.id = id;
+      this.parentId = parentId;
+      this.columnPath = columnPath;
+    }
+
+    public StructField field() {
+      return field;
+    }
+
+    public int id() {
+      return id;
     }
 
-    return columnIdToField;
+    public Integer parentId() {
+      return parentId;
+    }
+
+    public String columnPath() {
```

Comment:
Plan on using this in all the error messages, just didn't want to change all of those right now since it'll widen the changes since quite a few test assertions will need to change. This was always planned in a follow on (and I think it should be tackled along with making sure the actual exceptions we want are being thrown, not just IllegalArgumentExceptions) , but just including in the structure for now.